### DISCUSSION TITLE
lexer returns SyntaxError for i64::MIN literal

### DISCUSSION THREAD

**Discussion by hnakamur** [mitsuhiko/minijinja#351] [Q&A] [answered]
Currently the lexer returns SyntaxError for i64::MIN literal `{{ -9223372036854775808 }}`

I added it to `tests/lexer-inputs/literals.txt`
https://github.com/hnakamur/minijinja/commit/4cbcac8a0df5351e644e328b095c46f63c482127
and I got the following error.
```
thread 'test_lexer' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: SyntaxError, detail: "invalid integer" }', minijinja/tests/test_lexer.rs:17:41
```

The lexer parses an integer literal to i64 at
https://github.com/mitsuhiko/minijinja/blob/27413fbaeaeb218f081e4a39bc60e60bf6dc1902/minijinja/src/compiler/lexer.rs#L255

However the negative sign and 9223372036854775808  are tokenized separately, so 9223372036854775808 overflows since 9223372036854775808 > i64::MAX (= 9223372036854775807).

I think it would be nice for the lexer to handle a negative integer literal and handle i64::MIN literal, or document the current behavior that the available integer literal is between i64::MIN + 1 and i64::MAX.

**Comment by mitsuhiko** [comment, marked as answer]
I wanted to fix up some of those edge cases for a while but I was too lazy. The correct thing to do would be for the lexer to have to tokens: a `u64` for integers and a `u128` for large integers. Then `neg` (which is currently invoked for all negative integers) gets a special case for `170141183460469231731687303715884105728` which is the largest possible positive integer that can still be negated.  Independently of that, it's a bit ridiculous that the compiler emits opcodes for negations of numbers to begin with. It should really compile `-42` into just a constant rather than `Neg` + `LoadConst`.

I will fix this.

**Reply by mitsuhiko** [reply]
https://github.com/mitsuhiko/minijinja/pull/353

**Reply by hnakamur** [reply]
Thanks!
